### DISCUSSION TITLE
How can I check and prettify markdown source code

### DISCUSSION THREAD

**Discussion by bwl21** [thephpleague/commonmark#984] [Q&A]
I would like to 

* check the markdown source (e.g. prohibit tables)
* beautify the markdown source such that we have a standardized input format for easy diffing and maintenance.

**Comment by colinodell** [comment]
Hi there!

> check the markdown source (e.g. prohibit tables)

This could be done by parsing (but not rendering) the Markdown and checking [the generated AST](https://commonmark.thephpleague.com/2.4/customization/abstract-syntax-tree/) for the forbidden nodes.  For example (untested):

```php
$markdown = '...'; // todo

$environment = new Environment();
$environment->addExtension(new CommonMarkCoreExtension());
$environment->addExtension(new GithubFlavoredMarkdownExtension());

$parser = new MarkdownParser($environment);

$ast = $parser->parse($markdown);

foreach ($ast->iterator() as $node) {
    if ($node instanceof Table) {
        echo "Hey, tables aren't allowed!";
    }
}
```

> beautify the markdown source such that we have a standardized input format for easy diffing and maintenance.

We have an open feature request for this: #419.  I don't have an ETA on when this might be implemented - it could be a long time.  If you need something sooner I'd recommend looking at an alternative library like [markdownlint](https://github.com/DavidAnson/markdownlint)
